base/src/main/java/proguard/optimize/inline/IterativeInstructionVisitor.java

```diff
@@ -0,0 +1,57 @@
+package proguard.optimize.inline;
+
+import proguard.classfile.Clazz;
+import proguard.classfile.Method;
+import proguard.classfile.attribute.CodeAttribute;
+import proguard.classfile.instruction.Instruction;
+import proguard.classfile.instruction.InstructionFactory;
+import proguard.classfile.instruction.visitor.InstructionVisitor;
+
+/**
+ * This class allows the programmer to visit the instructions of a given code attribute but if the code is changed
+ * restart the iteration of the instructions so that the program can operate on these new instructions.
+ * <p>
+ * You just use instructionsAccept to start iterating over the instructions, you provide an instructionVisitor and it
+ * will visit all the instructions in the code attribute.
+ * <p>
+ * By using setCodeChanged you can tell this class that you changed the code, it will when fetching the next instruction
+ * jump back to the first one in the updated code attribute.
+ */
```

Comment:
This is commonly known as calculating a fixed point (constantly calling the same function on its output until the value no longer changes). I would rename this as something that has `FixedPoint` in the name.

I'm a bit confused by the API of this class. It is an InstructionVisitor, so you still have to do a lot of boilerplate to actually use this (check the changed field and manually invoke the visitor again)? Can you make this a visitor that visits code attributes instead so you can move that functionality inside this class?
Then the API should just be: `codeAttribute.accept(new FixedPointCodeAttributeVisitor(myInstructionVisitor))` where `myInstructionVisitor` can be a special subtype of `InstructionVisitor` that has the `hasCodeChanged()` method. This should probably also have a guard against infinite loops, if the visitor always returns that the code has changed.

Reply:
The naming is probably a bit confusing yes, because it's not really an instruction visitor, it's more of a code attribute visitor but without actually implementing `AttributeVisitor`. 

The functionality for re-visiting is already in this class, it happens by just setting the offset back to 0. The programmer just has to tell the class that the code has changed by using `setCodeChanged`. The reason `hasCodeChanged()` exists is for when you are doing longer operations so for example you have a for loop that goes over the arguments of a method and on some arguments it might make a change to the code attribute, however if we made a change to the code attribute in one iteration we can't have another change in the next iteration, we have to break out of the loop and not make any more changes until we revisit the code attribute again. So it exists for the case that you could theoretically do multiple edits in your instructionvisitor which you don't want, you first want to exit the instruction visitor and then come back again. But it it will not re-visit the code attirubte until the instruction visitor is done so it provides a way to early exit.

Reply:
The downside of having a custom instruction visitor type is that  things like `InstructionOpCodeFilter`can no longer be used for `myInstructionVisitor` you always have to make a custom visitor. Another option might be to pass an `InstructionVisitor` in the argument of this custom visitor so you can keep using existing visitors but I'm not sure how well that would work currently but I'll give it a try.